src/lib/array.jl

```diff
@@ -434,19 +434,33 @@ end
   end
 end
 
+# Matrix of pairwise difference quotients
+Base.@propagate_inbounds function _pairdiffquot(f, i, j, x, fx, dfx, d²fx = nothing)
+  i == j && return dfx[i]
+  Δx = x[i] - x[j]
+  T = real(eltype(x))
+  if abs(Δx) ≤ sqrt(eps(T))
```

Comment:
> That check is pessimistic when a second derivative is provided

Oh, good catch.

> Also that's relative errors, so actually it would make sense to be more careful as to what this should be compared to (ie possibly f' should make an appearance in that check).

I tried to work this out, and in the case where `f'(x)` alone is used to approximate the finite difference, then we would need `f''(x)/f'(x)` to compute the relative error. Likewise, we'd need `f'''(x)` to compute the error when we use `f''(x)`. If we assume the two derivatives are roughly on the same order on the order of 1, then we're fine ignoring their ratio. Otherwise, I don't see how to proceed. There's always approximating with `Δf'(x)/Δf(x)`, but that itself can be unstable.

Reply:
Unless someone has any ideas on how to include `f'(x)` in the check, I'm going to call this resolved.